velox/docs/functions/spark/aggregate.rst

```diff
@@ -80,6 +80,14 @@ General Aggregate Functions
 
     Returns the last non-null value of `x`.
 
+.. spark:function:: max(x) -> [same as x]
+
+    Returns the maximum value of ``x``.
+    ``x`` must be an orderable type.
+    Note: 1. Timestamp is in microsecond precision.
```

Comment:
@zhli1142015 Would you generate the docs and double check that they render nicely?

> 1. Timestamp is in microsecond precision.

I assume this applies to all Spark functions and operators. If that's the case, let's find a different place to document that to avoid repeating this for every function.

> 2. Map type is not orderable type in Spark.

This is also true in Presto and Velox. Notice that MapType::isOrderable is false.

> 3. Nested nulls are allowed within complex types.

I assume this applies to all Spark functions which handle NULL as value, e.g. implement DISTINCT semantics. If that's the case, let's find a different place to document that to avoid repeating this for every function.

Reply:
Yes, I actually duplicate them in another PR, could you please help to suggest the place where we can document them?
https://github.com/facebookincubator/velox/pull/10038

Reply:
@zhli1142015 Perhaps, we could add "Spark Types" section in https://facebookincubator.github.io/velox/develop/types.html We can then explain there that Spark operates on timestamps with "microsecond" precision, while Presto with "millisecond" precision. We could then maybe provide an example where min(ts) function truncates inputs. 

CC: @rui-mo

Reply:
This sounds good to me. Maybe we should change this in a different PR, @mbasmanova  and @rui-mo how do you think?
There are some types semantic difference between Spark and Presto in my mind, like 
1. Spark operates on timestamps with "microsecond" precision, while Presto with "millisecond" precision.
2. Nested nulls are handled as value in Spark (From @mbasmanova Presto has same behavior, but Velox Presto code actually fails the comparsion of nested nulls, we need to double confirm)
3. MapType is treated as non-comparable type in Spark, whild Velox Presto treats MapType as comparable type.

@rui-mo, do you know others difference?

Reply:
> MapType is treated as non-comparable type in Spark

Do you mean non-orderable? If so, all 3 systems are aligned.

Reply:
I mean non-comparable, https://github.com/facebookincubator/velox/pull/10038.
Looks Spark only has the concept ordering, if the data type is comparable then it means it's orderable also. For example in Spark CollectSet function and comparison functions,  Spark validate if the input types are orderable only. Map type is not orderable, so map type data can't be input of equalTo fucntions etc.

CollectSet function:
https://github.com/apache/spark/blob/d7f633a53495fa7de8898809d82447d484673b0d/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala#L179

Comparison function:
https://github.com/apache/spark/blob/d7f633a53495fa7de8898809d82447d484673b0d/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala#L985
But in Velox data type is comparable doesn't mean it's also orderable. like MapType may be comparable but not orderable.
Please correct me if I'm wrong. Thanks.

Reply:
Updated, thanks.

Reply:
> @rui-mo, do you know others difference?

I also notice below limitation for map type in Spark

> grouping/join/window partition keys cannot be map type